### PR TITLE
New ride flow

### PR DESCRIPTION
**Pull request by tyang98**
### Summary 

This pull request is implements the new designs for the ride flow. Users are able to now choose their locations, choose what days a ride can be repeated, and send the information to the backend. This pull request will most definitely be over the 1000 diff limit, but I believe there is no way to split up this PR without each component of the ride flow being connected to each other. Thus, this PR will not be merged **until we thoroughly test through everything.** 

- [x] implemented new Rider design (including most recent designs)
- [x] added recurring ride selection to ride flow
- [x] Create ride from the frontend and send to the backend.  
- [x] Filter custom locations 

### Test Plan 

I tested the ride flow through an emulator and with Postman.

### Notes 

-  Variation on the location displaying design
- No integration of maps

### Screenshots 

Screen Shot 2021-03-20 at 10 37 37 PM
Screen Shot 2021-03-20 at 10 37 45 PM
Screen Shot 2021-03-20 at 10 38 01 PM
Screen Shot 2021-03-20 at 10 38 31 PM
Screen Shot 2021-03-20 at 10 38 38 PM

### CODE REVIEW

**Comment by mg2239** [comment]
Is it possible to split up this PR by screen? So we have a PR for each screen (but no ride flow), then the last PR implements the flow for testing?

**Review by helenxuyang — COMMENTED** [review]
Overall really good progress, this is a TON of work!! Has a few style issues especially with the buttons, clunky behavior related to text fields, and a bug in locations when creating a second ride. Left some questions in the code as well. 

Here's a screen recording of me messing around with the flow, issues I identified are listed below. 
https://user-images.githubusercontent.com/60579411/112789683-a4559e00-902b-11eb-8e2e-8ee42b11df78.mp4

Visual 
- back button icon isn't centered and button is cut off on the bottom on some pages, next button overflows -> unsure of the problem here, but the widget trees for the buttons seem really complicated right now and could probably be simplified?
- on the date/time selection, the end date and drop-off time don't seem to be in line vertically -> probably need a GridView 
- on the review page, the end date and drop-off time don't seem to be in line vertically -> probably need a GridView
- text field decorations seem to flash on and off on taps, not sure what's going on here

Issues with text/keyboard
- tapping the text inputs causes the keyboard to come into view for a split second before the next page comes up -> this might get fixed automatically if we simplify the flow anyway
- due to the keyboard coming up, only one row of the autofill for locations is visible -> there is some way in Flutter to have everything shift up when the keyboard opens so that you have space, I think it's a Scaffold property? 
- the keyboard remains open after navigating back from the select location dates, and doesn't seem to be closeable so you have to just use the next button on the keyboard until you get to the end -> again, might get fixed automatically if we simplify the flow 
- the keyboard remains open after selecting dates/times and allows the user to type in these fields which shouldn't be allowed -> since the user should pretty much just use these as buttons, maybe surround with an IgnorePointer and then GestureDetector to use onTap to open the next page

Issues when creating a second ride
- locations seem to be carrying over -> clear when the flow is complete
- probably related to the location issue, the set location button doesn't appear even when there are rides in both fields

Other changes:
The locations should be changed so that the provider just fetches locations once when the app first starts and saves this into a variable that all the pages can access rather than fetching every time, which shows the loading symbol and makes unnecessary requests

Not approving yet especially if this PR will get split into multiple PRs anyway, and will want to review it again in more detail to make sure I'm not missing anything else

File: lib/pages/ride-flow/FlowWidgets.dart [hunk 1]
```diff
@@ -0,0 +1,171 @@
+import 'package:flutter/material.dart';
+
+class SelectionButton extends StatelessWidget {
+  final Widget page;
+  final String text;
+  final bool repeatPage;
+  final GestureTapCallback onPressed;
+
+  const SelectionButton(
+      {Key key, this.page, this.text, this.repeatPage, this.onPressed})
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/FlowWidgets.dart:10]: I think I initially wanted to do something with repeating a particular page in the flow.

File: lib/pages/ride-flow/FlowWidgets.dart [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: <Widget>[
+        Expanded(
+          child: new Container(
+            transform: Matrix4.translationValues(0.0, -15.0, 0.0),
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/FlowWidgets.dart:149]: Yeah it shifts the icon 15 pixels up. Without it, the distance between the icon and the upper bar is quite distant.

File: lib/pages/ride-flow/Request_Ride_Loc.dart [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+                          ),
+                        ]))),
+              ),
+            ],
+          ),
+        )));
+  }
+}
+
+class LocationInput extends StatefulWidget {
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Loc.dart:404]: Yes I agree, this should be a stateless widget

File: lib/pages/ride-flow/Review_Ride.dart [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+              children: <Widget>[
+                Text('Every', style: CarriageTheme.labelStyle)
+              ],
+            ),
+            SizedBox(height: 5),
+            Row(
+              mainAxisAlignment: MainAxisAlignment.start,
+              children: <Widget>[
+                Text(
+                    widget.ride.recurringDays != null
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Review_Ride.dart:165]: Not sure, the designs do not provide an instance of this case

File: lib/models/Location.dart [hunk 5]
```diff
@@ -1,39 +1,36 @@
...
-        data.map<Location>((e) => Location.fromJson(e)).toList();
-    return locations;
-  } else {
-    throw Exception('Failed to fetch locations.');
-  }
+  Map<String, dynamic> toJson() =>
+      {'id': id, 'name': name, 'address': address, 'info': info, 'tag': tag};
 }
 
 //Converts a list of locations [locations] to a Map containing location ids (key) to locations (value).
```
Comment (helenxuyang) [inline lib/models/Location.dart:36]: I know this wasn't changed here, but the locationsById function should probably be moved into LocationsProvider?

File: lib/pages/ride-flow/FlowWidgets.dart [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  Widget build(BuildContext context) {
+    return Expanded(
+      child: Align(
+          alignment: Alignment.bottomCenter,
+          child: Padding(
+              padding: const EdgeInsets.only(bottom: 30.0),
+              child: Row(
+                children: <Widget>[
+                  ButtonTheme(
+                    minWidth: MediaQuery.of(context).size.width * 0.1,
```
Comment (helenxuyang) [inline lib/pages/ride-flow/FlowWidgets.dart:101]: Button still looks a bit wide and the arrow isn't centered on my device, so I'm not sure if minWidth is the best thing to use here

File: lib/models/Location.dart [hunk 7]
```diff
@@ -1,39 +1,36 @@
...
-  if (response.statusCode == 200) {
-    String responseBody = response.body;
-    var data = jsonDecode(responseBody)['data'];
-    List<Location> locations =
-        data.map<Location>((e) => Location.fromJson(e)).toList();
-    return locations;
-  } else {
-    throw Exception('Failed to fetch locations.');
-  }
+  Map<String, dynamic> toJson() =>
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/models/Location.dart:32]: Oh i don't think it is necessary anymore. Beforehand, I was trying to send an entire location object to the backend.

File: lib/providers/RidesProvider.dart [hunk 8]
```diff
@@ -121,25 +122,34 @@ class RidesProvider with ChangeNotifier {
...
-      body: jsonEncode(<String, dynamic>{
-        'rider': riderProvider.info,
-        'startLocation': startLocation,
-        'endLocation': endLocation,
-        'startTime': startTime,
-        'requestedEndTime': endTime,
-      }),
-    );
+    Map<String, dynamic> request = <String, dynamic>{
+      'rider': riderProvider.info,
```
Comment (helenxuyang) [inline lib/providers/RidesProvider.dart]: riderProvider.info.id

**Review by helenxuyang — CHANGES_REQUESTED** [review]
Overall looking functional as a whole! I will update locations in a new PR because there's already so much going on here. Not approving yet because of the line that needs to be added for each TextFormField and the design stuff for the review ride (awkward for non-recurring rides) + ride confirmation (adding the button back) pages, but will approve after those. Awesome work Tony!!

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 9]
```diff
@@ -0,0 +1,699 @@
...
+      setState(() {
+        selectedDate = picked;
+        dateCtrl.text = format('$selectedDate'.split(' ')[0]);
+      });
+  }
+
+  Widget _buildPickupTimeField() {
+    return TextFormField(
+      controller: pickUpCtrl,
+      focusNode: AlwaysDisabledFocusNode(),
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:163]: done

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 10]
```diff
@@ -0,0 +1,699 @@
...
+      onTap: () => selectPickUpTime(context),
+      style: TextStyle(color: Colors.black, fontSize: 15),
+    );
+  }
+
+  Widget _buildDropOTimeField() {
+    return TextFormField(
+      controller: dropOffCtrl,
+      focusNode: AlwaysDisabledFocusNode(),
+      decoration: InputDecoration(
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:191]: done

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 11]
```diff
@@ -0,0 +1,699 @@
...
+                          TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
+                ],
+              ),
+              Form(
+                key: _formKey,
+                child: Column(
+                  children: <Widget>[
+                    TextFormField(
+                      controller: dateCtrl,
+                      focusNode: AlwaysDisabledFocusNode(),
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:262]: done

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 12]
```diff
@@ -0,0 +1,699 @@
...
+        ctrl.text = format('$date'.split(' ')[0]);
+      });
+  }
+
+  Widget _buildPickupTimeField() {
+    return Container(
+        width: MediaQuery.of(context).size.width / 3,
+        margin: EdgeInsets.only(left: 15.0),
+        child: TextFormField(
+          focusNode: AlwaysDisabledFocusNode(),
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:416]: done

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 13]
```diff
@@ -0,0 +1,699 @@
...
+                child: Column(
+                  children: <Widget>[
+                    Row(
+                      children: <Widget>[
+                        Container(
+                          width: MediaQuery.of(context).size.width / 3,
+                          margin: EdgeInsets.only(left: 15.0),
+                          child: TextFormField(
+                            controller: startDateCtrl,
+                            focusNode: AlwaysDisabledFocusNode(),
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:552]: done

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 14]
```diff
@@ -0,0 +1,699 @@
...
+                                _selectDate(context, startDate, startDateCtrl),
+                            style: TextStyle(color: Colors.black, fontSize: 15),
+                          ),
+                        ),
+                        SizedBox(width: 30),
+                        Container(
+                          width: MediaQuery.of(context).size.width / 3,
+                          margin: EdgeInsets.only(right: 15.0),
+                          child: TextFormField(
+                            focusNode: AlwaysDisabledFocusNode(),
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:585]: done

File: lib/pages/ride-flow/Request_Ride_Time.dart [hunk 15]
```diff
@@ -0,0 +1,699 @@
...
+                              labelText: 'End Date',
+                              labelStyle:
+                                  TextStyle(color: Colors.grey, fontSize: 17),
+                            ),
+                            validator: (input) {
+                              if (input.isNotEmpty &&
+                                  startDateCtrl.text != '' &&
+                                  DateFormat('MM/dd/yyyy')
+                                          .parse(startDateCtrl.text)
+                                          .difference(DateFormat('MM/dd/yyyy')
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Request_Ride_Time.dart:598]: thanks for the tip! fixed

File: lib/pages/ride-flow/Review_Ride.dart [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+                  crossAxisAlignment: CrossAxisAlignment.start,
+                  children: [
+                    Container(
+                      child: Column(
+                        crossAxisAlignment: CrossAxisAlignment.start,
+                        children: <Widget>[
+                          Text('End Date', style: CarriageTheme.labelStyle),
+                          SizedBox(height: 5),
+                          Text(
+                              widget.ride.endDate != null
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/Review_Ride.dart:121]: I agree, will remove

**Review by helenxuyang — APPROVED** [review]

File: lib/pages/ride-flow/FlowWidgets.dart [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+    );
+  }
+}
+
+class FlowBack extends StatelessWidget {
+  const FlowBack({Key key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Expanded(
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/FlowWidgets.dart:95]: I was having issues with positioning the widget

File: lib/pages/ride-flow/ToggleButton.dart [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+  final String text;
+  final List<bool> values;
+
+  @override
+  _ToggleButtonState createState() => new _ToggleButtonState();
+}
+
+class _ToggleButtonState extends State<ToggleButton> {
+  void _buttonChange() {
+    setState(() {
```
[the remark this answers is not in the document]
Comment (tyang98) [reply lib/pages/ride-flow/ToggleButton.dart:16]: I think setState is necessary

**Pull request merged**